src/main/java/com/linkedin/kafka/clients/producer/ExtensibleProducerRecord.java

```diff
@@ -0,0 +1,149 @@
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+
+/**
+ * <p>
+ * This class adds the capability to have key-value pairs, headers, which are delivered to the consumer.  The
+ * consumer must be using {@link com.linkedin.kafka.clients.consumer.LiKafkaConsumer}.  If headers are present and the
+ * consumer is not using
+ * {@link com.linkedin.kafka.clients.consumer.LiKafkaConsumer} then the consumer records will not be deserializable.
+ * True null values can not be sent in the presence of any headers.
+ * </p>
+ * <p>{@link HeaderUtils} contains suggestions for how to partition the header key into
+ * intervals.  Header keys must be non-null, non-zero length strings that can be encoded into a utf-8 string.  Headers
+ * live in a separate space from the underlying Kafka protocol.
+ * </p>
+ * <p> A header can be added to a producer record by calling {@link #header(String, byte[])}.  This implies that the
+ * user of the producer record should not modify the record after
+ * {@link com.linkedin.kafka.clients.producer.LiKafkaProducer#send} has been called.
+ * </p>
+ */
+public class ExtensibleProducerRecord<K, V> extends ProducerRecord<K, V> {
+
+  private Map<String, byte[]> headers;
+
+  /**
+   * Creates a record with a specified timestamp to be sent to a specified topic and partition
+   *
+   * @param topic The topic the record will be appended to
+   * @param partition The partition to which the record should be sent.  This may be null.
+   * @param timestamp The timestamp of the record.  This may be null.
+   * @param key The key that will be included in the record.  This may be null.
+   * @param value The record contents.  This may be null.
+   *
+   */
+  public ExtensibleProducerRecord(String topic, Integer partition, Long timestamp, K key, V value) {
+    super(topic, partition, timestamp, key, value);
+  }
+
+  /**
+   * Gets the value associated with the header key from this record.  This method is not thread safe.
+   * @param headerKey non-null, non-zero length string that can be encoded into a utf-8 string
+   * @return returns null if this record does not have headers of the header is not present
+   */
+  public byte[] header(String headerKey) {
+    HeaderUtils.validateHeaderKey(headerKey);
+
+    if (headers == null) {
+      return null;
+    }
+    return headers.get(headerKey);
+  }
+
+  /**
+   * Adds or updates the value associated with the header key.  This method is not thread safe.
+   * @param headerKey non-null, non-zero length string that can be encoded into a utf-8 string
+   * @param headerValue non-null
+   */
+  public void header(String headerKey, byte[] headerValue) {
+    HeaderUtils.validateHeaderKey(headerKey);
+
+    if (headerValue == null) {
+      throw new IllegalArgumentException("Header value must not be null.");
+    }
+
+    if (headers == null) {
+      headers = new TreeMap<>();
```

Comment:
When we had integer keys TreeMap had slightly better performance.  For String keys HashMap has slightly better performance.